### PR TITLE
Pointer communicator

### PR DESCRIPTION
**Pull request by RiccardoRossi**
this provides the framework for gathering generic data by the use of global pointers.

It takes ideas of PGAS (Partitioned Global Addres Space)  in providing a proxy to remote results.

@jcotela the "generic" serializer-based SendRecv funciton is implemented in the PointerCommunicator class. As agreed it should be moved to the data communicator.

### CODE REVIEW

**Comment by philbucher** [comment]
this seems like a cool feature :+1: 

You are using synchronous communication right?
Wouldn't it be smarter/more efficient to do this asynchronous?
=> I can imagine this to be used a lot in the future, and then refactoring will be painful

When I rewrote th MappingApp not so long ago I switched from sync to async and found that the implementation was even simpler

What do you think?

**Comment by RiccardoRossi** [comment]
@philbucher there is surely space for switching to async communications inside, nevertheless this is transparent to the user so i would leave it for later.

File: kratos/mpi/utilities/pointer_communicator.h
Comment (jcotela) [inline kratos/mpi/utilities/pointer_communicator.h]: Please update or remove

File: kratos/utilities/global_pointer_utilities.h
Comment (jcotela) [inline kratos/utilities/global_pointer_utilities.h]: "for now" :man_facepalming:

File: kratos/utilities/global_pointer_utilities.h
Comment (jcotela) [inline kratos/utilities/global_pointer_utilities.h]: Most likely can be `const`

File: kratos/mpi/utilities/pointer_communicator.h
Comment (jcotela) [inline kratos/mpi/utilities/pointer_communicator.h]: Can/should the argument of the function to be applied remotely be const (`const GlobalPointer<TPointerDataType>&` or maybe `GlobalPointer<const TPointerDataType>&`?). In either case, I assume that we do not want to allow the global pointer operation to modify the remote object.

File: kratos/mpi/utilities/pointer_communicator.h
Comment (jcotela) [inline kratos/mpi/utilities/pointer_communicator.h]: The format here is inconsistent. I assume it should be `mCurrentRank`, `mNonLocalData`, `mUserFunction`.

**Review by jcotela — COMMENTED** [review]
Overall I am positive about this PR. How should we proceed regarding the things that need to be integrated into the data communicator(and eventually the wrappers for separate Send/Recv calls?).

File: kratos/mpi/utilities/pointer_communicator.h
Comment (RiccardoRossi) [reply kratos/mpi/utilities/pointer_communicator.h]: why not? imagine you want to mark a flag as visited...

**Comment by philbucher** [comment]
@RiccardoRossi I am folowing the global pointers stuff since the beginning but I still don't get how to use them :)
Do you have an example somewhere? I think it would be quite helpful for the mpi-mapping stuff

**Comment by RiccardoRossi** [comment]
take a look at the test example. i will try anyway to post an rxplanation
here asap

**Comment by philbucher** [comment]
Thanks, I hadn't seen it.

We should speak abt this next week, I think this will make my life much easier :D

**Comment by RiccardoRossi** [comment]
here goes a quick note about the intended use of global pointers. What is implemented right now i**s a way to simplify gathering of arbitrary data**. At a later step a function for sending data will also be implemented, nevertheless we (@KratosMultiphysics/technical-committee) we did not yet converge on a good interface so sending data is left as a second step.

so back to what we have right now.

let's consider a loop to take the average of neighbours nodes for each node.
In serial it looks something like

~~~cpp
for(auto& node : model_part.Nodes)
    double Tavg = 0;
    const auto& neighbours = node.GetValue(NODAL_NEIGHBOURS);
    for(auto& neighb : neighbours)
         Tavg += neighb.FastGetSolutionStepValue(TEMPERATURE);
    node.SetValue(AVERAGE_TEMPERATURE, Tavg/(neigbours.size());
}
~~~

this loop is obviously not possible in mpi. (we do not have all the neighbours on a given process unless we add a lot of ghost nodes)

Let's now consider we use global pointers instead of weak_pointers for the neighbours. IN THE SERIAL CASE the previous loop would look very similar, something like 

~~~cpp
for(auto& node : model_part.Nodes)
    double Tavg = 0;
    const auto& global_pointers_to_neighbours = node.GetValue(GP_NODAL_NEIGHBOURS);
    for(auto& gp : global_pointers_to_neighbours)
         Tavg += gp->FastGetSolutionStepValue(TEMPERATURE);
    node.SetValue(AVERAGE_TEMPERATURE, Tavg/(global_pointers_to_neighbours.size());
}
~~~

Since global pointers are pointers that are only valid **in the memory space of the owner processor**
this works ok on a single process, but still fails in mpi, since the global neighbours can belong to different mpi processes.

The current PR makes an attempt to simplify dealing with this sort of situations:
in order to work in mpi we would have to modify the code as follows:


~~~cpp
//first of all collect all the neighbours into a list
GlobalPointerVector< Node<3 > > gp_list;

for(auto& node : model_part.Nodes)
    for(auto& gp : global_pointers_to_neighbours)
        gp_list.push_back(gp);

//now create the pointer communicator
PointerCommunicator pointer_comm(DefaultDataCommunicator, gp_list); //here we prepare the communication

//define a function to access to the data
result_proxy = pointer_comm.Apply<double>(
            [](GlobalPointer<Node<3> >& gp) -> double
            {
                  return gp->FastGetSolutionStepValue(TEMPERATURE);
            }
      );  //here all communications happen !!

for(auto& node : model_part.Nodes)
    double Tavg = 0;
    const auto& global_pointers_to_neighbours = node.GetValue(GP_NODAL_NEIGHBOURS);
    for(auto& gp : global_pointers_to_neighbours)
         Tavg += result_proxy.Get(gp); //here we get the result of the application of the function onto gp. This is so BOTH for LOCAL GPs and for REMOTE GPs
    node.SetValue(AVERAGE_TEMPERATURE, Tavg/(global_pointers_to_neighbours.size());
}
~~~

the code is of course a little more verbose, nevertheless all the communication is hidden in the Apply function (and in the constructor), and the use if still more or less similar to the original serial code.

Aside of this, the code can be made to be still efficient in serial. The point here is that the function "get" of the "result_proxy" is actually **executing the function directly if the globalpointer is local**. This means that no remote data at all is needed (and hence computed) if the communicator returns IsDistributed() -> false.
This also means that additional storage can be avoided completely in the serial case.
In the practice this means that we can further optimize the code to have no overhead in the serial case as follows:

~~~cpp
//first of all collect all the neighbours into a list ----> NOT ACTUALLY NEEDED IN SERIAL
GlobalPointerVector< Node<3 > > gp_list;

if(DefaultDataCommunicator->IsDistributed() == true) //an emty gp_list is enough in the serial case
    for(auto& node : model_part.Nodes)
        for(auto& gp : global_pointers_to_neighbours)
             gp_list.push_back(gp);

//now create the pointer communicator --> NOTHING IS ACTUALLY DONE IN THE SERIAL CASE!
PointerCommunicator pointer_comm(DefaultDataCommunicator, gp_list); //here we prepare the communication

//define a function to access to the data ----> NO COMMUNICATION HAPPENS IN THE SERIAL CASE
//we only store the access function within result_proxy and we will use it on every "Get" call
result_proxy = pointer_comm.Apply<double>(
            [](GlobalPointer<Node<3> >& gp) -> double
            {
                  return gp->FastGetSolutionStepValue(TEMPERATURE);
            }
      );  //here all communications happen !!

for(auto& node : model_part.Nodes)
    double Tavg = 0;
    const auto& global_pointers_to_neighbours = node.GetValue(GP_NODAL_NEIGHBOURS);
    for(auto& gp : global_pointers_to_neighbours)
         Tavg += result_proxy.Get(gp); //here we get the result of the application of the function onto gp. This is so BOTH for LOCAL GPs and for REMOTE GPs
    node.SetValue(AVERAGE_TEMPERATURE, Tavg/(global_pointers_to_neighbours.size());
}
~~~
with such modification the code is almost as fast in serial as the original code: a function call by function pointer is done on every global pointer, but the lock that was needed for the weak pointer is avoided.

**Comment by RiccardoRossi** [comment]
just to add that "Apply" is templated in the result_type. 
this implies that the function passed as parameter (here i use a lambda function, but it could be anything) 
can return any type of object, and this object **will be transparently returned to the caller object** as long as the object is serializable and does not contain pointers.

File: kratos/mpi/utilities/pointer_communicator.h
Comment (RiccardoRossi) [reply kratos/mpi/utilities/pointer_communicator.h]: done

File: kratos/mpi/utilities/pointer_communicator.h
Comment (RiccardoRossi) [reply kratos/mpi/utilities/pointer_communicator.h]: done

File: kratos/utilities/global_pointer_utilities.h
Comment (RiccardoRossi) [reply kratos/utilities/global_pointer_utilities.h]: done

**Comment by RiccardoRossi** [comment]
@jcotela i think i addressed all of your comments.

regarding the SendRecv i would merge this, and once it is in move the templated SendRecv to the data communicator. 
Once that is done i would remove the implementation inside the pointer_communicator and inside the global_pointer_utilities.

other TODOs are related to missing Send and Recv separate functions (no hurry to implement).
Also generic Scatter/gather/broadcast are not present 

as a final comment, i now added PointerComm to the MPI core **however pointer_comm** perfectly works in serial mode with the default serial communicator.
This means that it would probably be better to have it in the includes subdir of the core

**Comment by pooyan-dadvand** [comment]
@RiccardoRossi Thanks for your detail explanation!

Now you are passing a constructed gp_list to the constructor. Maybe we can open an alternative way of creating the communicator with a function which creates the list:
~~~cpp
//Create a pointer communicator with function --> NOTHING IS ACTUALLY DONE IN THE SERIAL CASE!
PointerCommunicator pointer_comm(DefaultDataCommunicator, 
          [model_part.Nodes](GlobalPointerVector< Node<3 > >& GpList) {
              for(auto& node : model_part.Nodes)
                  for(auto& gp : global_pointers_to_neighbours)
                       GpList.push_back(gp);
          }
);  //here we prepare the communication

//define a function to access to the data ----> NO COMMUNICATION HAPPENS IN THE SERIAL CASE
//we only store the access function within result_proxy and we will use it on every "Get" call
result_proxy = pointer_comm.Apply<double>(
            [](GlobalPointer<Node<3> >& gp) -> double
            {
                  return gp->FastGetSolutionStepValue(TEMPERATURE);
            }
      );  //here all communications happen !!

for(auto& node : model_part.Nodes)
    double Tavg = 0;
    const auto& global_pointers_to_neighbours = node.GetValue(GP_NODAL_NEIGHBOURS);
    for(auto& gp : global_pointers_to_neighbours)
         Tavg += result_proxy.Get(gp); //here we get the result of the application of the function onto gp. This is so BOTH for LOCAL GPs and for REMOTE GPs
    node.SetValue(AVERAGE_TEMPERATURE, Tavg/(global_pointers_to_neighbours.size());
}
~~~

and having some predefined functions for usual use cases to reduce the code:
~~~cpp
//Create a pointer communicator with function --> NOTHING IS ACTUALLY DONE IN THE SERIAL CASE!
PointerCommunicator pointer_comm(DefaultDataCommunicator, CollectNeighboursGlobalPointers( model_part.Nodes));  //here we prepare the communication

//define a function to access to the data ----> NO COMMUNICATION HAPPENS IN THE SERIAL CASE
//we only store the access function within result_proxy and we will use it on every "Get" call
result_proxy = pointer_comm.Apply<double>(GetVariableFromGlobalPointer(TEMPERATURE));  //here all communications happen !!

for(auto& node : model_part.Nodes)
    double Tavg = 0;
    const auto& global_pointers_to_neighbours = node.GetValue(GP_NODAL_NEIGHBOURS);
    for(auto& gp : global_pointers_to_neighbours)
         Tavg += result_proxy.Get(gp); //here we get the result of the application of the function onto gp. This is so BOTH for LOCAL GPs and for REMOTE GPs
    node.SetValue(AVERAGE_TEMPERATURE, Tavg/(global_pointers_to_neighbours.size());
}
~~~


This way we are hiding the gp_list considering it as implementation detail. 

Nevertheless I should check the implementation to see if is viable

**Comment by RiccardoRossi** [comment]
@pooyan-dadvand i have no problem in adding the constructor you propose, i think it would look something like

~~~cpp
GlobalPointerCommunicator(DataCommunicator& rComm, 
                                              std::function< GlobalPointersVector< TPointerDataType >(void)& 
         InputFunction )
        :
        mrDataCommunicator(rComm)
    {
        
         if(mrDataCommunicator.IsDistributed())
        {
             auto gp_list = InputFunction() //note that we do not need to pass the list, we can obtain it from the function
            AddPointers(gp_list.begin(),gp_list.end());

            //compute communication plan
            std::vector<int> send_list;
            send_list.reserve( mNonLocalPointers.size() );
            for(auto& it : mNonLocalPointers)
                send_list.push_back( it.first );

             std::sort(send_list.begin(), send_list.end());
            mColors = MPIColoringUtilities::ComputeCommunicationScheduling(send_list, mrDataCommunicator);
        }
    }
~~~

the difference from your proposal is that i am not passing gp_list to the function, but rather getting it as a return value.

what i don't like about this is (and also about my implementation of the proxy) is that we cannot pass a functor (i mean a class with an operator (...) mimicking the interface).
The only (minor) problem is that the TInputFunctionType should be a template parameter of the constructor if we want to allow that. 
I am actually very willing to add this template parameter both for the new constructor as well as for the apply in order to allow the use of functors.

**Comment by pooyan-dadvand** [comment]
> the difference from your proposal is that i am not passing gp_list to the function, but rather getting it as a return value.

:+1: 

> I am actually very willing to add this template parameter both for the new constructor as well as for the apply in order to allow the use of functors.

I think it would be good idea!

File: kratos/CMakeLists.txt
Comment (philbucher) [inline kratos/CMakeLists.txt:104]: ouch...

now we have an MPI-core, why not putting it there?
If I understand correctly this is only needed for MPI right?

File: kratos/python/add_processes_to_python.cpp
Comment (philbucher) [inline kratos/python/add_processes_to_python.cpp]: I don't think this should be in "add_processes_to_python.cpp"

File: kratos/CMakeLists.txt
Comment (RiccardoRossi) [reply kratos/CMakeLists.txt:104]: the mpi_coloring_utilities.cpp does not have any dependency on mpi (it only depends on DataCommunicator) and is included in the pointer_communicator which can be used both in serial and in parallel code (so when the mpi core is not loaded)

this means that as a utility it needs to go in the core and in the mpi_core

File: kratos/python/add_processes_to_python.cpp
Comment (RiccardoRossi) [reply kratos/python/add_processes_to_python.cpp]: I see your point, nevertheless i am not really sure of where to put it

File: kratos/CMakeLists.txt
Comment (jcotela) [reply kratos/CMakeLists.txt:104]: I did not understand what do you mean by 
> this means that as a utility it needs to go in the core and in the mpi_core

Do we need to add it to both libraries?

Also, if it has no relation to mpi, why is it called `mpi_....`?

Comment (RiccardoRossi) [reply kratos/CMakeLists.txt:104]: the phrase should have been:
in the core and NOT in the mpi_core.

just think: pointer_comm can be used when you have no mpi right? it includes the mpi_coloring_utilities.h , so this file needs to be available when you compile the pointer_comm.

maybe we could rename it as 

mpi_coloring_utilities -> communication_coloring_utilities

**Comment by RiccardoRossi** [comment]
the basic idea of the latest changes is to make #4729 independent on this so that we have time to improve the current PR while pushing through #4729 and #4779

File: kratos/CMakeLists.txt
Comment (jcotela) [reply kratos/CMakeLists.txt:104]: :+1: for the rename.

Comment (RiccardoRossi) [reply kratos/CMakeLists.txt:104]: @jcotela i would prefer to make the renaming in a different PR as soon as this is merged. 

the point is that here i already moved the mpi_coloring_utilities, and i would really like to avoid further messing up the same file.

File: kratos/utilities/pointer_communicator.h
Comment (roigcarlo) [inline kratos/utilities/pointer_communicator.h]: Shouldn't we remove this?

Comment (RiccardoRossi) [reply kratos/utilities/pointer_communicator.h]: please don't, if something goes wrong this is fundamental to debug it, and it takes a while to rewrite it

**Comment by RiccardoRossi** [comment]
@philbucher after merging the latest master it does work without explicitly importing mpi.

File: kratos/CMakeLists.txt
Comment (RiccardoRossi) [reply kratos/CMakeLists.txt:104]: (note that i am referring to renaming the name of the file)

**Comment by pooyan-dadvand** [comment]
@KratosMultiphysics/technical-committee approves this as first iteration and comments and suggestions are welcomed!

**Review by roigcarlo — DISMISSED** [review]

**Review by roigcarlo — APPROVED** [review]
:+1:

**Pull request merged**
